lib/AST/Decl.cpp

```diff
@@ -5062,6 +5072,45 @@ bool FuncDecl::isDeferBody() const {
   return getName() == getASTContext().getIdentifier("$defer");
 }
 
+bool FuncDecl::isXCTestUnitTest() const {
+  if (!hasName())
+    return false;
+
+  // A 'test candidate' is:
+  // 1. An instance method...
+  if (!isInstanceMember())
+    return false;
+
+  // 2. ...on a class or extension (not a struct) subclass of XCTestCase...
+  auto parentClass = getDeclContext()->getAsClassOrClassExtensionContext();
+  if (!parentClass)
+    return false;
+  if (!parentClass->isXCTestCaseSubclass())
+    return false;
+
+  // 3. ...that returns void...
+  Type retTy = getResultInterfaceType();
+  if (retTy && !retTy->isVoid())
+    return false;
+
+  // 4. ...takes no parameters...
+  if (getParameterLists().size() != 2)
+    return false;
+  if (getParameterList(1)->size() != 0)
+    return false;
+
+  // 5. ...is of at least 'internal' accessibility (unless we can use
+  //    Objective-C reflection)...
+  ASTContext &ctx = getASTContext();
+  if (!ctx.LangOpts.EnableObjCInterop &&
+      (getFormalAccess() < Accessibility::Internal ||
+       parentClass->getFormalAccess() < Accessibility::Internal))
+    return false;
+
+  // 6. ...and starts with "test".
```

Comment:
I think this is correct as-is. XCTest.framework includes `- (void)testfoo {}` / `func testfoo() {}` when it gathers test methods at runtime.